halo2_proofs/src/plonk/circuit.rs

```diff
@@ -1083,6 +1181,64 @@ impl<F: Field> ConstraintSystem<F> {
         index
     }
 
+    /// Add a dynamic lookup argument for some input expressions and table columns.
+    ///
+    /// `table_map` returns a map between input expressions and the table columns
+    /// they need to match.
+    ///
+    /// `table` must contain all table columns used in table_map.
+    #[cfg(feature = "unstable-dynamic-lookups")]
+    pub fn lookup_dynamic(
+        &mut self,
+        table: &DynamicTable,
+        table_map: impl FnOnce(
+            &mut VirtualCells<'_, F>,
+        ) -> (Selector, Vec<(Expression<F>, Column<Any>)>),
+    ) -> usize
+    where
+        F: From<u64>,
+    {
+        let mut cells = VirtualCells::new(self);
+        let (selector, table_map) = table_map(&mut cells);
+        let selector = cells.query_selector(selector);
+
+        let mut table_map: Vec<_> = table_map
+            .into_iter()
+            .map(|(input, table_col)| {
+                if !cells.meta.dynamic_tables[table.0].columns.contains(&table_col) {
+                    panic!(
+                        "{:?} does not contain {:?}. Try adding this column to the dynamic table.",
+                        table,
+                        table_col
+                    );
+                }
+
+                if selector.contains_simple_selector() {
+                    panic!("selector expression containing simple selector supplied to lookup argument");
+                }
+
+                // TODO is this needed
```

Comment:
```suggestion
```
Yes this is needed. A simple selector column doesn't actually exist after optimization.